### PR TITLE
 Add an option to export/import the preferences. 

### PR DESCRIPTION
**Pull request by TheBolliwood**
Implemented a simple JSON export and import (#86).
I tried to keep #30 in mind that a sync could be added later easily (#30).

### CODE REVIEW

**Comment by TheBolliwood** [comment]
The migration of the preferences from older versions is currently missing. I'll add that later.

File: src/background/runtime.js [hunk 1]
```diff
@@ -44,6 +44,9 @@ class Runtime {
         this.permissions.notifications = true;
       }
       this.storage.local.preferences = message.payload.preferences;
+      if(message.payload.migrate) {
+        this.migration.onUpdate({previousVersion: message.payload.previousVersion});
+      }
       await this.storage.persist();
```
Comment (stoically) [inline src/background/runtime.js]: The migration should be called after the storage.persist.

Nit: whitespace after if

File: src/ui/shared.js [hunk 2]
```diff
@@ -443,19 +448,15 @@ window.importPreferences = async (file) => {
...
   reader.readAsText(file, 'UTF-8');
   reader.onload = async (e) => {
-    let preferences;
-    try {
-      preferences = JSON.parse(e.target.result);
-    } catch (error) {
-      // eslint-disable-next-line no-console
-      console.log('error while importing preferences', error);
-      showError('Error while importing preferences!');
-    }
```
Comment (stoically) [inline src/ui/shared.js]: Why did you remove the catch / showError?

File: src/ui/shared.js [hunk 3]
```diff
@@ -431,7 +431,12 @@ window.exportPreferencesButton = async () => {
 
 window.exportPreferences = async () => {
   const storage = await browser.storage.local.get('preferences');
-  return JSON.stringify(storage.preferences, null, 2);
+  const exportedPreferences = {
+    version: browser.runtime.getManifest().version,
+    date: Date.now(),
```
Comment (stoically) [inline src/ui/shared.js]: I'd rather use `new Date` instead of the timestamp. We probably won't get bitten by the [Year 2038 Problem](https://en.wikipedia.org/wiki/Year_2038_problem), but who knows.

Comment (stoically) [reply src/ui/shared.js]: So I just learned that javascript uses [floats instead of integers](https://ecma-international.org/ecma-262/5.1/#sec-8.5) and thus isn't prone to the mentioned problem, so you can scratch that and leave as is if you prefer.

File: src/ui/shared.js (hunk 2 above)
Comment (TheBolliwood) [reply src/ui/shared.js]: I merged the two try/catch statements into one.

Comment (stoically) [reply src/ui/shared.js]: Ah, makes sense. Thanks.

File: src/background/runtime.js (hunk 1 above)
Comment (TheBolliwood) [reply src/background/runtime.js]: Thanks!

File: src/background/runtime.js [hunk 4]
```diff
@@ -44,10 +44,10 @@ class Runtime {
         this.permissions.notifications = true;
       }
       this.storage.local.preferences = message.payload.preferences;
-      if(message.payload.migrate) {
+      await this.storage.persist();
+      if (message.payload.migrate) {
         this.migration.onUpdate({previousVersion: message.payload.previousVersion});
```
Comment (stoically) [inline src/background/runtime.js]: One last thing and we're good to merge. The `onUpdate` call should have an `await`, so that the migration is guaranteed to finish before the `initialize` in the options UI is called (it `await`s the `sendMessage` there).

**Comment by grenzor** [comment]
One thing I'd like to ask, is it possible to append time/date to the filename so it's something like `TC_settings_2019-04-04_12.00.00.json`? Adding the time/date will make it easier to sort and identify specific backups.

File: src/background/runtime.js (hunk 4 above)
Comment (TheBolliwood) [reply src/background/runtime.js]: Oh, right. I forgot that. Thank you.

**Comment by TheBolliwood** [comment]
@grenzor Thanks for your suggestion!

**Comment by stoically** [comment]
Published with version 0.96. @TheBolliwood Thanks!
